### PR TITLE
Add LDAP auth module

### PR DESCRIPTION
**Pull request by brodo**
This PR adds a auth module for [LDAP](https://ldap.com/) servers. There are some people who have rolled their own implementations of this (you can find them by searching for 'ldap' in the issues) so there is definitely a need for it. The PR adds a dependency on the [ldapjs](http://ldapjs.org/) package. 

I tried to make it as generic as possible, so it *should* work with all kinds of LDAP servers and configurations. This means that it takes a lot of config options, but I'll open a PR in the `docs` repo shortly where I describe how to set it up.

### CODE REVIEW

**Comment by brodo** [comment]
I don't know why Codecov thinks coverage in files which I did not touch has decreased. 😐

File: package.json [hunk 1]
```diff
@@ -38,6 +38,7 @@
     "graphql-upload": "8.1.0",
     "intersect": "1.0.1",
     "jsonwebtoken": "8.5.1",
+    "ldapjs": "^1.0.2",
```
Comment (dplewis) [inline package.json]: Can you pin the package? remove the `^`.

File: src/Adapters/Auth/ldap.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      ) {
+        const filter = options.groupFilter.replace(/{{id}}/gi, authData.id);
+        const opts = {
+          scope: 'sub',
+          filter: filter,
+        };
+        let found = false;
+        client.search(options.suffix, opts, (searchError, res) => {
+          if (searchError) {
+            throw new Parse.Error(
```
Comment (dplewis) [inline src/Adapters/Auth/ldap.js]: Can you add a test for this error?

File: src/Adapters/Auth/ldap.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  return (
+    typeof options === 'object' &&
+    typeof options.suffix === 'string' &&
+    typeof options.url === 'string' &&
+    options.url.startsWith('ldap://')
+  );
+}
+
+function validateAppId() {
+  return Promise.resolve();
```
Comment (dplewis) [inline src/Adapters/Auth/ldap.js]: Can you add your module [here](https://github.com/parse-community/parse-server/blob/master/spec/AuthenticationAdapters.spec.js#L44) for testing or a test case to ensure `validateAppId` exists?

File: package.json (hunk 1 above)
Comment (brodo) [reply package.json]: done!

File: src/Adapters/Auth/ldap.js (hunk 2 above)
Comment (brodo) [reply src/Adapters/Auth/ldap.js]: Added a test and found out that I forgot to destroy the client when this error happens...

File: src/Adapters/Auth/ldap.js (hunk 3 above)
Comment (brodo) [reply src/Adapters/Auth/ldap.js]: I added a test case to my spec which calls it and checks if it resolves the returned promise.

File: src/Adapters/Auth/ldap.js (hunk 2 above)
Comment (dplewis) [reply src/Adapters/Auth/ldap.js]: Can you just double check if you are destroying / unbinding the client in the correct places? If everything is good we can get this merged.

File: src/Adapters/Auth/ldap.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+function searchForGroup(client, options, id, resolve, reject) {
+  const filter = options.groupFilter.replace(/{{id}}/gi, id);
+  const opts = {
+    scope: 'sub',
+    filter: filter,
+  };
+  let found = false;
+  client.search(options.suffix, opts, (searchError, res) => {
+    if (searchError) {
+      client.unbind();
```
Comment (dplewis) [inline src/Adapters/Auth/ldap.js]: Can you write a test case for this? LDAP group search failed error

Comment (brodo) [reply src/Adapters/Auth/ldap.js]: The test 'Should fail if the LDAP server encounters an error while searching' is now testing this case.

**Comment by brodo** [comment]
> Can you just double check if you are destroying / unbinding the client in the correct places? If everything is good we can get this merged.

There was one more place where I forgot the destroy/unbind. Should be all fixed now.

**Pull request merged**